lib/galaxy/tools/__init__.py

```diff
@@ -3563,6 +3563,67 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
         )
 
 
+class HarmonizeTool(DatabaseOperationTool):
+    tool_type = "harmonize_list"
+    require_terminal_states = False
+    require_dataset_ok = False
+
+    def produce_outputs(self, trans, out_data, output_collections, incoming, history, **kwds):
+        # Get the 2 input collections
+        hdca1 = incoming["input1"]
+        hdca2 = incoming["input2"]
+        # Get the elements of both collections
+        elements1 = hdca1.collection.elements
+        elements2 = hdca2.collection.elements
+        # Put elements in dictionary with identifiers:
+        old_elements1_dict = {}
+        for element in elements1:
+            old_elements1_dict[element.element_identifier] = element
+        old_elements2_dict = {}
+        for element in elements2:
+            old_elements2_dict[element.element_identifier] = element
+        # Get the list of final identifiers
+        final_sorted_identifiers = [element.element_identifier for element in elements1
+                                    if element.element_identifier in old_elements2_dict]
+        # Raise Exception if it is empty
+        if len(final_sorted_identifiers) == 0:
+            # Create empty collections:
+            output_collections.create_collection(
+                next(iter(self.outputs.values())), "output1", elements={}, propagate_hda_tags=False
+            )
+            output_collections.create_collection(
+                next(iter(self.outputs.values())), "output2", elements={}, propagate_hda_tags=False
+            )
+        # Create outputs:
+        new_elements1 = {}
+        new_elements2 = {}
+        for identifier in final_sorted_identifiers:
+            # First collection output
```

Comment:
The logic for first and second collection output generation can be factored out into a simple helper method: the only difference in lines 3602-3607 and 3609-3614 is the 2 dictionaries (old and new elements) - you could just pass them as arguments and call the helper twice.

Reply:
thanks for the tip